src/lib.rs

```diff
@@ -0,0 +1,182 @@
+use colored::Colorize;
+use std::{
+    fmt,
+    time::{Duration, Instant},
+};
+
+const RUNTIME_THRESHOLD: Duration = Duration::from_millis(5);
+
+pub struct FunctionBenchmark {
+    pub runtime: Option<Duration>,
+    memory_size: Option<u32>,
+    stack_size: Option<u32>,
+    start: Option<Instant>,
+}
```

Comment:
I think this structure is a little too flexible. It'd be more clear if a function benchmark always represented the benchmarking results of a function, rather than occasionally holding intermediate values.

I think that timing a code block is standard enough that we can just use it directly in main. This also eliminates the need to `unwrap` the start value.

Reply:
I find that having the logic outside of the main makes it more readable and less coupled, and that having the possibility to instantiate the FunctionBenchmark struct outside the code that executes it makes it more testable. I understand that it is more verbose and the simpler would work well, but I think it would be a lot harder to test. Another reason of why I did it this way is because I would want to simplify the main by moving the engine logic in another function. This way I would create the benchmark struct in the main and pass it in the function that controls the engine.

Reply:
Caveat here is that this is all a nit (which means a non-critical comment, doesn't need to prevent a PR from being merged)

We require that the function benchmark is calculated by a `start` and `stop` call, which only has one realistic implementation of time.now/elapsed. Given that, these two blocks are functionally the same:
```
benchmark.start
... code ...
benchmark.end
```
```
time = now
... code ...
benchmark = Benchmark::new(time.elapsed)
```
If I consider a world where we add a second engine implementation that calls out to Shopify servers to get a more accurate measurement by running the script on Shopify infrastructure, then what this benchmark measures would no longer be accurate. Or, if the engine performed a series of time consuming operations around the execution, like performing `wasm-opt`, this would also not work. 

Additionally, each engine will likely have a bespoke way of measuring memory. Then we'd have a `Benchmark` object measuring time, and the engine separately measuring memory, and later we'd combine them.

I think we need to tweak the ownership here towards the engine, which is currently embedded in main. Then we'll be able to get the full benefits of the indirection. Later we can extract the engine, which gives us readability.

Reply:
Additionally on readability, just because it's an interesting topic. There's benefits to this kind of abstraction, but there's also a cost:
- It deviates from the standard library. If I'm a new developer to the repository already familiar with Rust, I need to go learn what `Benchmark` is and when `start` and `stop` need to be called. When I see a timer and `time.elapsed` calls I know exactly how that works without following any links.
- It introduces a new name into the namespace of actors. In one model, there's an engine and it gets run, returning a benchmark value object. In the other, we have an Engine, and a Benchmark, both of which contain logic that interoperates.

Reply:
Thank you for this great feedback! Learned a lot 😃 

I'll implement this change and merge